### PR TITLE
Implement connection inactivity timeout in C++

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR implements the connection inactivity timeout in C++.

A connection is inactive (as opposed to idle) when it's active (a confusing terminology, see below) and has no outstanding invocation or dispatch.

This PR adds a connection _dispatchCount to count dispatches. It's reasonably easy to manage. I originally planned to include incoming heartbeats in this dispatch count when the heartbeat callback is not null, but did not. Sending or receiving a heartbeat has no effect on the inactivity timeout.

For "invocation count", I am only counting two-way invocations (the size of the _asyncRequests map). Sending any invocation cancels the inactivity timeout timer task. For oneway invocations, we consider the invocation complete as soon as it's queued--we don't wait until it's actually sent. It would be nicer to wait until it's sent but the code is too complicated.

> active vs inactive

Rather than use the term "inactive", this PR uses "at rest". A connection is at rest when it's active and has no outstanding invocation or dispatch.

### CODE REVIEW

File: cpp/src/Ice/ConnectionI.cpp [hunk 1]
```diff
@@ -2310,6 +2382,19 @@ Ice::ConnectionI::idleCheck(
...
+void
+Ice::ConnectionI::inactivityCheck() noexcept
+{
+    // Called by the InactivityTimerTask.
+    std::lock_guard lock(_mutex);
+    if (isAtRest())
+    {
+        setState(
+            StateClosing,
+            make_exception_ptr(CloseConnectionException{__FILE__, __LINE__, "connection closed by inactivity timer"}));
```
Comment (bernardnormier) [inline cpp/src/Ice/ConnectionI.cpp:2394]: I think it's the correct exception to use even though I find its name odd. I'd rather not create yet another local exception.

Assuming we keep this exception here, we'll need to update its doc comment.

File: cpp/src/Ice/ConnectionI.h [hunk 2]
```diff
@@ -348,6 +376,9 @@ namespace Ice
         // The number of user calls currently executed by the thread-pool (servant dispatch, invocation response, ...)
         int _upcallCount;
 
+        // The number of outstanding dispatches, include heartbeat callback "dispatches".
```
Comment (externl) [inline cpp/src/Ice/ConnectionI.h:379]:
```suggestion
        // The number of outstanding dispatches, including heartbeat callback "dispatches".
```

File: cpp/src/Ice/ConnectionI.cpp [hunk 3]
```diff
@@ -3115,6 +3206,12 @@ Ice::ConnectionI::parseMessage(int32_t& upcallCount, function<bool(InputStream&)
                         return false; // the upcall will be completed once the dispatch is done.
                     };
                     ++upcallCount;
+
```
Comment (externl) [inline cpp/src/Ice/ConnectionI.cpp:3209]: Presumably this `upcallCount` still gets decremented correctly later?

File: cpp/src/Ice/ConnectionI.cpp (hunk 1 above)
Comment (externl) [inline cpp/src/Ice/ConnectionI.cpp:2394]: The user won't know about an inactivity timer. They just want to know why it's closed. 
```suggestion
            make_exception_ptr(CloseConnectionException{__FILE__, __LINE__, "connection closed due to inactivity"}));
```

We could maybe even add the time. `connection closed due to inactivity after X seconds`

**Review by externl — APPROVED** [review]
Looks good. Hard to verify that _dispatchCount is. Maybe we could add some asserts on destruction or something to ensure it's back to 0?

File: cpp/src/Ice/ConnectionI.cpp [hunk 4]
```diff
@@ -2310,6 +2382,19 @@ Ice::ConnectionI::idleCheck(
     // else, nothing to do
 }
 
+void
+Ice::ConnectionI::inactivityCheck() noexcept
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/ConnectionI.cpp:2386]: It's consistent with the "idle check" function name, idleCheck.

File: cpp/src/Ice/ConnectionI.h [hunk 5]
```diff
@@ -348,6 +376,9 @@ namespace Ice
         // The number of user calls currently executed by the thread-pool (servant dispatch, invocation response, ...)
         int _upcallCount;
 
+        // The number of outstanding dispatches, including heartbeat callback "dispatches".
```
Comment (bentoi) [inline cpp/src/Ice/ConnectionI.h:379]: The implementation says otherwise, `_dispatchCount` isn't increased for heartbeats upcalls.

**Review by bentoi — CHANGES_REQUESTED** [review]
It looks like a connection that started reading a message (but didn't fully read the payload) is considered inactive. And as you pointed out, it connection which is writing a oneway request will also be considered inactive. Actually, I think it's worse: a connection which is writing a response is also considered inactive.

As a result, the server will close the connection in these cases. I don't think it's good idea and it doesn't match IceRPC's behavior.

I would look into checking for `_readHeader == false` (if no header is read, there's no pending message for read) and `_writeStream.b.empty()` (if the write buffer is empty, there's no message being written).

It also doesn't look like the timer is scheduled after successful connection establishment (a connection can be established without sending a request by using `proxy->ice_getConnection()`) .

**Comment by bernardnormier** [comment]
> It looks like a connection that started reading a message (but didn't fully read the payload) is considered inactive. 

This is both accurate and the correct behavior. Merely receiving a protocol frame does not make a connection "not at rest". This frame must be an (incoming) request for a connection to transition from "at rest" to "not at rest" - and cancel the inactivity timer task.

I think having this check at the end of "case requestMsg:" and before we dispatch the request is the right spot. Where do you suggest to put this check for incoming requests? 

> And as you pointed out, it connection which is writing a oneway request will also be considered inactive.

This is not that bad. When we send the oneway request, we cancel the inactivity timer task. So the sending of this oneway request has to take over inactivity timeout (by default 5min) for the connection to be shutdown by the inactivity timeout. 

> Actually, I think it's worse: a connection which is writing a response is also considered inactive.

Same here: the worst case scenario is we schedule the inactivity timer task right after writing the response. So with the default config, the sending of the response has to take over 5min for the connection to be shutdown by the inactivity timeout.

> 
> As a result, the server will close the connection in these cases. I don't think it's good idea and it doesn't match IceRPC's behavior.

I would also prefer to wait under after a oneway request is sent / a response it sent to "decrement" the "dispatch+invocation" count. It's just much harder to do with Ice, hence this compromise. 

> 
> I would look into checking for `_readHeader == false` (if no header is read, there's no pending message for read) and `_writeStream.b.empty()` (if the write buffer is empty, there's no message being written).
> 

What about you fix this behavior in a follow-up PR?

> It also doesn't look like the timer is scheduled after successful connection establishment (a connection can be established without sending a request by using `proxy->ice_getConnection()`) .

The timer is scheduled when the connection becomes Active. This has nothing to do with connection establishment, since a just established connection is in the Holding state.

I think it's important for the inactivity timeout to have an effect only when the connection state is Active (as per this PR); my understanding is we can't perform a graceful shutdown for a connection in the Holding state and the inactivity timeout is all about graceful shutdown.

File: cpp/src/Ice/ConnectionI.h (hunk 5 above)
Comment (bernardnormier) [reply cpp/src/Ice/ConnectionI.h:379]: Fixed comment.

**Comment by bernardnormier** [comment]
> We already created a new exception so I would add a new exception instead of re-use this exception which is a protocol exception.

How many exceptions do you want to define for a gracefully closed connection?

We already have:
 - ConnectionCloseException
that's what you get when the peer initiated the graceful shutdown (is that really important?)
 - ConnectionManuallyClosedException with graceful = true
that's when you call `close` on your Connection

In IceRPC, we have IceRpcError.ConnectionClosedByPeer (=ConnectionCloseException) and IceRpcError.ConnectionAborted.

File: cpp/src/Ice/ConnectionI.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Ice/ConnectionI.cpp]: I updated the message.

**Comment by bernardnormier** [comment]
> > We already created a new exception so I would add a new exception instead of re-use this exception which is a protocol exception.

I added a new exception, `ConnectionClosedException`, which takes a message. It's currently C++-only but needs to be ported to other languages (in a follow-up PR).

I think we should rework these exceptions in a follow-up PR. For example:
 - CloseConnectionException remains as-is and derives from ProtocolException.. It's named after the CloseConnection ice protocol frame and corresponds to the ConnectionClosedByPeer error code in IceRPC. 
 - ConnectionClosedException: graceful closure with a message
 - ConnectionAbortedException: non-graceful closure with a message

And we remove ConnectionIdleException (replaced by ConnectionClosedException) and ConnectionManuallyClosedException (replaced by ConnectionClosedException / ConnectionAbortedException).

**Comment by bentoi** [comment]
> This is both accurate and the correct behavior. Merely receiving a protocol frame does not make a connection "not at rest". This frame must be an (incoming) request for a connection to transition from "at rest" to "not at rest" - and cancel the inactivity timer task.

Correct me if I'm wrong but this is the case for the IceRPC ice's protocol implementation but not for the icerpc protocol implementation. In the ice protocol implementation we actually schedule the inactive timer task only when done with dispatching. It doesn't look like it's rescheduled upon fully reading a reply. Sounds like a bug?

> I think having this check at the end of "case requestMsg:" and before we dispatch the request is the right spot. Where do you suggest to put this check for incoming requests?

Nothing to suggest because I think it's a mistake to have a specific behavior for  request messages.

> So with the default config, the sending of the response has to take over 5min for the connection to be shutdown by the inactivity timeout.

And with a non default config where the application wants to close inactive connections quickly? How do you prevent the connection from being closed while there are several large responses or large oneway requests being queued for sending?

> However, I believe your idea is a non-starter. Heartbeats should have no effect on the inactivity timeout, and _readHeader / _writeStream.b.empty() are most likely affected by heartbeats.

Good point. Let's suppose we can exclude the sending/receive of a heartbeat from the inactivity check. Do you agree that we should only consider the connection inactive if there's no data being received/sent and no invocations/dispatches in progress? Wouldn't this match IceRPC's icerpc protocol implementation?

The connection validation message has no payload and should always be read/sent in one shot (very unlikely that a transport sends 13 bytes in several steps!). So I think it's probably doable:
- we cancel the inactive timer as soon as we start reading a message which is not a connection validation message
- we cancel the inactive timer as soon as we send a message which is not a connection validation message
- we reschedule the timer once all these conditions are true:
  - `_writeStream.b.empty() == true`, there's no more data to send
  - `_readHeader == true`, there's no more message to read and we are waiting for a new message
  - `(invocations + dipatches count) == 0`, no more invocations or dispatches in progress

File: cpp/src/Ice/ConnectionI.cpp [hunk 6]
```diff
@@ -2193,6 +2267,13 @@ Ice::ConnectionI::setState(State state)
             setState(StateClosed, current_exception());
         }
     }
+
+    if (isAtRest())
```
Comment (bentoi) [inline cpp/src/Ice/ConnectionI.cpp:2271]: Shouldn't we check for `_state == StateActive` here?

Comment (bernardnormier) [reply cpp/src/Ice/ConnectionI.cpp:2271]:
```cpp
bool isAtRest() const noexcept
{
    return _state == StateActive && _dispatchCount == 0 && _asyncRequests.empty();
}
```

**Comment by bernardnormier** [comment]
> Correct me if I'm wrong but this is the case for the IceRPC ice's protocol implementation but not for the icerpc protocol implementation. In the ice protocol implementation we actually schedule the inactive timer task only when done with dispatching. It doesn't look like it's rescheduled upon fully reading a reply. Sounds like a bug?

The IceRPC code is much simpler. In IceProtocolConnection.cs, we call DecrementDispatchInvocationCount when an invocation is complete (e.g. after reading the reply):
https://github.com/icerpc/icerpc-csharp/blob/57995b702436169798fc3fe5fd513e6f09591397/src/IceRpc/Internal/IceProtocolConnection.cs#L453

This method reschedules the inactivity timeout timer when the count reaches 0:
https://github.com/icerpc/icerpc-csharp/blob/57995b702436169798fc3fe5fd513e6f09591397/src/IceRpc/Internal/IceProtocolConnection.cs#L987 

I don't see any bug here.

> Good point. Let's suppose we can exclude the sending/receive of a heartbeat from the inactivity check. Do you agree that we should only consider the connection inactive if there's no data being received/sent and no invocations/dispatches in progress? Wouldn't this match IceRPC's icerpc protocol implementation?

Yes, I agree. However, figuring out when a request/response is actually sent is difficult with the Ice code. That's why I didn't do it.
 
> The connection validation message has no payload and should always be read/sent in one shot (very unlikely that a transport sends 13 bytes in several steps!).

You meant 14 bytes?

> So I think it's probably doable:
> 
> * we cancel the inactive timer as soon as we start reading a message which is not a connection validation message
> * we cancel the inactive timer as soon as we send a message which is not a connection validation message

Ok.

> * we reschedule the timer once all these conditions are true:
>   
>   * `_writeStream.b.empty() == true`, there's no more data to send

That's because when we send a heartbeat - and just a heartbeat - the sending is synchronous and _writeStream.b.empty() can't be true? In other words, `_writeStream.b.empty() == true` means we're sending asynchronously something else than a heartbeat?

>   * `_readHeader == true`, there's no more message to read and we are waiting for a new message
>   * `(invocations + dipatches count) == 0`, no more invocations or dispatches in progress

where invocations = _asyncRequests.size()?
If not, how do you count invocations in progress?

**Comment by bernardnormier** [comment]
> we reschedule the timer once all these conditions are true:
_writeStream.b.empty() == true, there's no more data to send
_readHeader == true, there's no more message to read and we are waiting for a new message
(invocations + dipatches count) == 0, no more invocations or dispatches in progress

With this proposal, how do we get notified this condition may have become true and it's now time to schedule the inactivity timer task? I'd rather not have a background thread that checks for this condition from time to time.

In my current code, the trigger is when _dispatchCount is decremented (and therefore can reach 0) and a reply is received (and therefore _asyncRequests can become empty).

**Comment by bernardnormier** [comment]
@bentoi will open a PR to improve this implementation.

**Pull request merged**